Add ToBloq adapter class and to_bloq function to convert PennyLane circuits and operators to Qualtran bloqs (#7197)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

  [(#7197)](https://github.com/PennyLaneAI/pennylane/pull/7197)
  
  :func:`qml.to_bloq <pennylane.to_bloq>` translates PennyLane operators into equivalent [Qualtran bloqs](https://qualtran.readthedocs.io/en/latest/bloqs/index.html#bloqs-library). It 
  requires one input and takes in two optional inputs:
  * `circuit`: an initialized PennyLane operator or a QNode
  * `map_ops`: whether or not if operators are mapped to a Qualtran Bloq or wrapped as a `ToBloq`
  * `custom_mapping`: dictionary to specify a mapping between a PennyLane operator and a Qualtran Bloq
  
  The following example converts a PennyLane Operator into a Qualtran Bloq:

  ```python
  import pennylane as qml
  from qualtran.drawing import get_musical_score_data, draw_musical_score, show_bloq

  control_wires = [2, 3]
  estimation_wires = [4, 5, 6, 7, 8, 9]
  
  H = -0.4 * qml.Z(0) + 0.3 * qml.Z(1) + 0.4 * qml.Z(0) @ qml.Z(1)
  
  op = qml.QuantumPhaseEstimation(
      qml.Qubitization(H, control_wires), estimation_wires=estimation_wires
  )
  
  cbloq = qml.to_bloq(op).decompose_bloq()
  fig, ax = draw_musical_score(get_musical_score_data(cbloq))

>   show_bloq(cbloq)

Author: I'm not getting any errors on my end. It is the same code as what was in the demo.

  # Let's define a custom mapping instead

  from qualtran.bloqs.phase_estimation import LPResourceState
  from qualtran.bloqs.phase_estimation.text_book_qpe import TextbookQPE

  custom_map = {
    op: TextbookQPE(
        unitary=qml.to_bloq(qml.Qubitization(H, control_wires)), 
        ctrl_state_prep=LPResourceState(len(estimation_wires))
    )
  }

  cbloq = qml.to_bloq(op, map_ops=True, custom_mapping=custom_map).decompose_bloq()
  fig, ax = draw_musical_score(get_musical_score_data(cbloq))
  show_bloq(cbloq)
  ```

  Alternatively, rather than map directly to a Qualtran Bloq, we can preserve the original
  PennyLane decomposition by setting map_ops to False.

  ```python
  op_wrapped_as_bloq = qml.to_bloq(op, map_ops=False)
  cbloq = op_wrapped_as_bloq.decompose_bloq()
  fig, ax = draw_musical_score(get_musical_score_data(cbloq))

>   show_bloq(cbloq)

Reviewer: Shouldn't we be drawing the figure and axis? If not can we just not save the return here?

Author: I see this commonly done in Qualtran. When you call draw_musicial_score it will draw the musical score as well as give you some returns so that you can potentially modify the figure. 

e.g.
```
fig, ax = draw_musical_score(get_musical_score_data(cbloq))
fig.some_setting = some_value
```

In this case, draw_musical_score will draw the musical score, and show_bloq will also create an image, so you get two images total.

Author: (It does look a little confusing though. Maybe I should just remove the fig, ax part?

  [(#7355)](https://github.com/PennyLaneAI/pennylane/pull/7355)
  [(#7586)](https://github.com/PennyLaneAI/pennylane/pull/7586)

<h4>Qualtran Integration 🔗</h4>

* It's now possible to convert PennyLane operators to [Qualtran](https://qualtran.readthedocs.io/en/latest/) bloqs with the new :func:`qml.to_bloq <pennylane.to_bloq>` function. 
  [(#7197)](https://github.com/PennyLaneAI/pennylane/pull/7197)
  
  :func:`qml.to_bloq <pennylane.to_bloq>` translates PennyLane operators into equivalent [Qualtran bloqs](https://qualtran.readthedocs.io/en/latest/bloqs/index.html#bloqs-library). It 
  requires one input and takes in two optional inputs:
  * op (Operation): an initialized PennyLane operator to be wrapped as a Qualtran Bloq.

>   * map_ops (bool): Whether or not if the operations are mapped to a Qualtran Bloq or wrapped as a ``ToBloq``. Default is True.

Reviewer 2: Maybe something like this?
```suggestion
  * map_ops (bool): Whether to map operations to a Qualtran Bloq. Operations are wrapped as a ``ToBloq`` when False. Default is True.
```
